File: Algorithms/Medium/24.Swap-Nodes-In-Pairs/SwapNodesInPairs.go
package medium

type ListNode struct {
	Val  int
	Next *ListNode
}

// Time complexity: O(n)
// Space complexity: O(1)
func swapPairs(head *ListNode) *ListNode {
	dummy := &ListNode{Val: 0, Next: head}
	prev, curr := dummy, head

	for curr != nil && curr.Next != nil {
		next_next := curr.Next.Next
		next := curr.Next

		curr.Next = next_next
		next.Next = curr
		prev.Next = next

		prev = curr
		curr = next_next
	}

	return dummy.Next
}

File: Algorithms/Medium/24.Swap-Nodes-In-Pairs/SwapNodesInPairs.py
from typing import Optional


class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


# Time complexity: O(n)
# Space complexity: O(1)
class Solution:
    def swapPairs(self, head: Optional[ListNode]) -> Optional[ListNode]:
        dummy = ListNode(next=head)
        prev, curr = dummy, head

        while curr and curr.next:
            nxt = curr.next
            nxt_nxt = curr.next.next

            nxt.next = curr
            prev.next = nxt
            curr.next = nxt_nxt

            prev = curr
            curr = nxt_nxt

        return dummy.next
